Make the anonymiser more tolerant of non-unique primary keys and patients with invalid MPIDs

martinburchell opened the pull request:
Make the anonymiser more tolerant of non-unique primary keys and patients with invalid MPIDs. Tested on the CPFT server where this was originally observed.

@RudolfCardinal should this be configurable?

Also:
* Fix a typo in the installer workflow error handling routine (only seen when the workflow failed)
* Fix the Dockerfile, which was failing to install fuzzy due to setuptools incompatibility. We now ensure that setuptools is more recent
* A couple of security fixes (Django, cryptography)
* Removal of redundant test `odbc_user.ini`, which was renamed `odbc_user_container.ini` in e4eade85f2ac8c1a8bb9f124e92577735c2e0955

RudolfCardinal reviewed (commented):
Looks good. Two questions about transactions:
* At the DatabaseError trap: if patient 1 is inserted, and patient 2 fails, will the rollback then erase both? (If so then we'd need to make the admindb transaction more explicit, but I can't remember where the implicit commit is here.)
* I presume the other one (trapping IntegrityError) checks/raises within a transaction, rather than waiting for a commit, and so is OK?

Re configurability: mm, debatable. I think it's unlikely that users would want a great deal of granularity here. But maybe an overall "tolerate_database_errors" flag would be useful? I'm not sure; what do you think?

martinburchell commented:
@RudolfCardinal I've added a test to make sure that only the problematic patient is skipped. This made me realise that the approach we use for database testing fails when a transaction is rolled back, because the pytest fixtures run each test in a transaction in order to restore the database after each run. So I've made alternative "slow" pytest fixtures that instead of starting and rolling back a transaction, create and drop the tables for the secret database.

One thing you might want to check: I've removed the commit after the patient record is initially created with just the PID, which had the comment "prompt commit after insert operations, to ensure no locks". (See cec55b08f0139cfd7bfd6a76653effd237de198b). Assuming the MPID is valid, it will get committed right after all of the scrubber stuff has been initialised but leaving that commit in could end up with a Patient without an MPID if it is later discarded.

I'll leave the configuration option for now. We can always put it in later if there is a need for it.

RudolfCardinal reviewed (commented):
Thanks! I presume the ``IntegrityError`` doesn't mess up preceding inserts in the transaction. But could the ``DatabaseError`` around ``patient = Patient(pid)`` mess up other (inc. even other patient) records in the ``adminsession`` transaction? If so, would a way be to wrap this insertion within a specific sub-transaction? Or, since I see these may be becoming deprecated, by committing the admindb session just before trying this?

martinburchell commented:
> Thanks! I presume the `IntegrityError` doesn't mess up preceding inserts in the transaction. But could the `DatabaseError` around `patient = Patient(pid)` mess up other (inc. even other patient) records in the `adminsession` transaction? If so, would a way be to wrap this insertion within a specific sub-transaction? Or, since I see these may be becoming deprecated, by committing the admindb session just before trying this?

Well it certainly can't hurt. I've added a commit as you suggest.
